raft.go

```diff
@@ -1479,11 +1488,32 @@ func (r *Raft) requestVote(rpc RPC, req *RequestVoteRequest) {
 	// check the LeadershipTransfer flag is set. Usually votes are rejected if
 	// there is a known leader. But if the leader initiated a leadership transfer,
 	// vote!
-	candidate := r.trans.DecodePeer(req.Candidate)
-	if leader := r.Leader(); leader != "" && leader != candidate && !req.LeadershipTransfer {
+	var candidate ServerAddress
+	var candidateBytes []byte
+	if req.ProtocolVersion > 3 {
+		candidate = r.trans.DecodePeer(req.Addr)
+		candidateBytes = req.Addr
+	} else {
+		candidate = r.trans.DecodePeer(req.Candidate)
+		candidateBytes = req.Candidate
+	}
+
+	// Prior to version 4 the peer ID is not part of `RequestVoteRequest`,
+	// We assume that the peer is part of the configuration and skip this check
+	if req.ProtocolVersion > 3 {
+		candidateID := ServerID(req.ID)
+		// if the Servers list is empty that mean the cluster is very likely trying to bootstrap,
+		// Grant the vote
+		if len(r.configurations.latest.Servers) > 0 && !inConfig(r.configurations.latest, candidateID) {
+			r.logger.Warn("rejecting vote request since voter not part of the configuration",
+				"from", candidate)
+			return
+		}
+	}
+	if leaderAddr, leaderID := r.Leader(); leaderAddr != "" && leaderAddr != candidate && !req.LeadershipTransfer {
 		r.logger.Warn("rejecting vote request since we have a leader",
 			"from", candidate,
-			"leader", leader)
+			"leader (", leaderAddr, string(leaderID), ")")
```

Comment:
This might render ok for the string formatted logs, but I think this is going to look really strange in JSON output.
```suggestion
			"leader", leaderAddr,
			"leaderID", string(leaderID))
```
The varargs is supposed to be a set of key/value pairs, so I think this should be :

Reply:
yes I fixed some of those, this one slipped.